docs/configuration-guide.md
# Configuration Guide

The confd configuration file is written in [TOML](https://github.com/mojombo/toml)
and loaded from `/etc/confd/confd.toml` by default. You can specify the config file via the `-config-file` command line flag.

> Note: You can use confd without a configuration file. See [Command Line Flags](https://github.com/kelseyhightower/confd/blob/master/docs/command-line-flags.md).

Optional:

* `backend` (string) - The backend to use. ("etcd")
* `client_cakeys` (string) - The client CA key file.
* `client_cert` (string) - The client cert file.
* `client_key` (string) - The client key file.
* `confdir` (string) - The path to confd configs. ("/etc/confd")
* `interval` (int) - The backend polling interval in seconds. (600)
* `log-level` (string) - level which confd should log messages ("info")
* `nodes` (array of strings) - List of backend nodes. (["http://127.0.0.1:4001"])
* `noop` (bool) - Enable noop mode. Process all template resources; skip target update.
* `prefix` (string) - The string to prefix to keys. ("/")
* `scheme` (string) - The backend URI scheme. ("http" or "https")
* `srv_domain` (string) - The name of the resource record.
* `srv_record` (string) - The SRV record to search for backends nodes.
* `sync-only` (bool) - sync without check_cmd and reload_cmd.
* `watch` (bool) - Enable watch support.

Single backend example:

```TOML
confdir = "/etc/confd"
log-level = "debug"
interval = 600
noop = false

[[backend]]
id = "default"
type = "etcd"
client_cert = "/etc/confd/ssl/client.crt"
client_key = "/etc/confd/ssl/client.key"
nodes = [
  "http://127.0.0.1:4001",
]
prefix = "/production"
scheme = "https"
srv_domain = "etcd.example.com"
```

*Note* that there must **always** be at least one backend with the ID of "default"

Multiple backend example:

You must use a toml config file to be able to define multiple backends

```TOML
confdir = "/etc/confd"
log-level = "debug"
interval = 600
noop = false

[[backend]]
id = "default"
type = "etcd"
client_cert = "/etc/confd/ssl/client.crt"
client_key = "/etc/confd/ssl/client.key"
nodes = [
  "http://127.0.0.1:4001",
]
prefix = "/production"
scheme = "https"
srv_domain = "etcd.example.com"

[[backend]]
id = "env_vars"
type = "env"
prefix = "/production"
```


config.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/BurntSushi/toml"
	"github.com/kelseyhightower/confd/backends"
	"github.com/kelseyhightower/confd/log"
	"github.com/kelseyhightower/confd/resource/template"
)

var (
	configFile        = ""
	defaultConfigFile = "/etc/confd/confd.toml"
	authToken         string
	authType          string
	backend           string
	basicAuth         bool
	clientCaKeys      string
	clientCert        string
	clientKey         string
	confdir           string
	config            Config // holds the global confd config.
	interval          int
	keepStageFile     bool
	logLevel          string
	nodes             Nodes
	noop              bool
	onetime           bool
	prefix            string
	printVersion      bool
	scheme            string
	srvDomain         string
	srvRecord         string
	syncOnly          bool
	table             string
	templateConfig    template.Config
	backendsConfig    backends.Config
	username          string
	password          string
	watch             bool
	appID             string
	userID            string
)

// Backend defines a backend for confd
type Backend struct {
  ID           string   `toml:"id"`
  Type         string   `toml:"type"`
	Prefix       string   `toml:"prefix"`
	ClientCaKeys string   `toml:"client_cakeys"`
	ClientCert   string   `toml:"client_cert"`
	ClientKey    string   `toml:"client_key"`
	BasicAuth    bool     `toml:"basic_auth"`
	BackendNodes []string `toml:"nodes"`
	AuthToken    string   `toml:"auth_token"`
	AuthType     string   `toml:"auth_type"`
	SRVDomain    string   `toml:"srv_domain"`
	SRVRecord    string   `toml:"srv_record"`
	Scheme       string   `toml:"scheme"`
	Table        string   `toml:"table"`
	Username     string   `toml:"username"`
	Password     string   `toml:"password"`
	AppID        string   `toml:"app_id"`
	UserID       string   `toml:"user_id"`
}

// A Config structure is used to configure confd.
type Config struct {
	ConfDir      string    `toml:"confdir"`
	Noop         bool      `toml:"noop"`
	LogLevel     string    `toml:"log-level"`
	SyncOnly     bool      `toml:"sync-only"`
	Backend      []Backend `toml:"backend"`
	Watch        bool      `toml:"watch"`
	Interval     int       `toml:"interval"`
}

func init() {
	flag.StringVar(&authToken, "auth-token", "", "Auth bearer token to use")
	flag.StringVar(&backend, "backend", "etcd", "backend to use")
	flag.BoolVar(&basicAuth, "basic-auth", false, "Use Basic Auth to authenticate (only used with -backend=etcd)")
	flag.StringVar(&clientCaKeys, "client-ca-keys", "", "client ca keys")
	flag.StringVar(&clientCert, "client-cert", "", "the client cert")
	flag.StringVar(&clientKey, "client-key", "", "the client key")
	flag.StringVar(&confdir, "confdir", "/etc/confd", "confd conf directory")
	flag.StringVar(&configFile, "config-file", "", "the confd config file")
	flag.IntVar(&interval, "interval", 600, "backend polling interval")
	flag.BoolVar(&keepStageFile, "keep-stage-file", false, "keep staged files")
	flag.StringVar(&logLevel, "log-level", "", "level which confd should log messages")
	flag.Var(&nodes, "node", "list of backend nodes")
	flag.BoolVar(&noop, "noop", false, "only show pending changes")
	flag.BoolVar(&onetime, "onetime", false, "run once and exit")
	flag.StringVar(&prefix, "prefix", "", "key path prefix")
	flag.BoolVar(&printVersion, "version", false, "print version and exit")
	flag.StringVar(&scheme, "scheme", "http", "the backend URI scheme for nodes retrieved from DNS SRV records (http or https)")
	flag.StringVar(&srvDomain, "srv-domain", "", "the name of the resource record")
	flag.StringVar(&srvRecord, "srv-record", "", "the SRV record to search for backends nodes. Example: _etcd-client._tcp.example.com")
	flag.BoolVar(&syncOnly, "sync-only", false, "sync without check_cmd and reload_cmd")
	flag.StringVar(&authType, "auth-type", "", "Vault auth backend type to use (only used with -backend=vault)")
	flag.StringVar(&appID, "app-id", "", "Vault app-id to use with the app-id backend (only used with -backend=vault and auth-type=app-id)")
	flag.StringVar(&userID, "user-id", "", "Vault user-id to use with the app-id backend (only used with -backend=value and auth-type=app-id)")
	flag.StringVar(&table, "table", "", "the name of the DynamoDB table (only used with -backend=dynamodb)")
	flag.StringVar(&username, "username", "", "the username to authenticate as (only used with vault and etcd backends)")
	flag.StringVar(&password, "password", "", "the password to authenticate with (only used with vault and etcd backends)")
	flag.BoolVar(&watch, "watch", false, "enable watch support")
}

// initConfig initializes the confd configuration by first setting defaults,
// then overriding settings from the confd config file, then overriding
// settings from environment variables, and finally overriding
// settings from flags set on the command line.
// It returns an error if any.
func initConfig() error {
	if configFile == "" {
		if _, err := os.Stat(defaultConfigFile); !os.IsNotExist(err) {
			configFile = defaultConfigFile
		}
	}
	// Set defaults.
	config = Config{
		Backend: []Backend{
			Backend{
				ID: "default",
				Type: "etcd",
				Prefix:   "",
				Scheme:   "http",
			},
		},
		ConfDir:  "/etc/confd",
		Interval: 600,
	}
	// Update config from the TOML configuration file.
	if configFile == "" {
		log.Debug("Skipping confd config file.")
	} else {

		log.Debug("Loading " + configFile)
		configBytes, err := ioutil.ReadFile(configFile)
		if err != nil {
			return err
		}
		_, err = toml.Decode(string(configBytes), &config)
		if err != nil {
			return err
		}
	}

	// Update config from environment variables.
	processEnv()

	// Update config from commandline flags.
	processFlags()

	if config.LogLevel != "" {
		log.SetLevel(config.LogLevel)
	}

	if config.Backend[0].SRVDomain != "" && config.Backend[0].SRVRecord == "" {
		config.Backend[0].SRVRecord = fmt.Sprintf("_%s._tcp.%s.",
			config.Backend[0].Type, config.Backend[0].SRVDomain)
	}

	// Update BackendNodes from SRV records.
	if config.Backend[0].Type != "env" && config.Backend[0].SRVRecord != "" {
		log.Info("SRV record set to " + config.Backend[0].SRVRecord)
		srvNodes, err := getBackendNodesFromSRV(config.Backend[0].SRVRecord, config.Backend[0].Scheme)
		if err != nil {
			return errors.New("Cannot get nodes from SRV records " + err.Error())
		}
		config.Backend[0].BackendNodes = srvNodes
	}
	if len(config.Backend[0].BackendNodes) == 0 {
		switch config.Backend[0].Type {
		case "consul":
			config.Backend[0].BackendNodes = []string{"127.0.0.1:8500"}
		case "etcd":
			peerstr := os.Getenv("ETCDCTL_PEERS")
			if len(peerstr) > 0 {
				config.Backend[0].BackendNodes = strings.Split(peerstr, ",")
			} else {
				config.Backend[0].BackendNodes = []string{"http://127.0.0.1:4001"}
			}
		case "redis":
			config.Backend[0].BackendNodes = []string{"127.0.0.1:6379"}
		case "vault":
			config.Backend[0].BackendNodes = []string{"http://127.0.0.1:8200"}
		case "zookeeper":
			config.Backend[0].BackendNodes = []string{"127.0.0.1:2181"}
		}
	}
	// Initialize the storage client
	log.Info("Backend set to " + config.Backend[0].Type)

	if config.Watch {
		unsupportedBackends := map[string]bool{
			"redis":    true,
			"dynamodb": true,
			"rancher":  true,
		}

		if unsupportedBackends[config.Backend[0].Type] {
			log.Info(fmt.Sprintf("Watch is not supported for backend %s. Exiting...", config.Backend[0].Type))
			os.Exit(1)
		}
	}

	if config.Backend[0].Type == "dynamodb" && config.Backend[0].Table == "" {
		return errors.New("No DynamoDB table configured")
	}

	backendsConfig = backends.Config{
		AuthToken:    config.Backend[0].AuthToken,
		AuthType:     config.Backend[0].AuthType,
		Backend:      config.Backend[0].Type,
		BasicAuth:    config.Backend[0].BasicAuth,
		ClientCaKeys: config.Backend[0].ClientCaKeys,
		ClientCert:   config.Backend[0].ClientCert,
		ClientKey:    config.Backend[0].ClientKey,
		BackendNodes: config.Backend[0].BackendNodes,
		Password:     config.Backend[0].Password,
		Scheme:       config.Backend[0].Scheme,
		Table:        config.Backend[0].Table,
		Username:     config.Backend[0].Username,
		AppID:        config.Backend[0].AppID,
		UserID:       config.Backend[0].UserID,
	}
	// Template configuration.
	templateConfig = template.Config{
		ConfDir:       config.ConfDir,
		ConfigDir:     filepath.Join(config.ConfDir, "conf.d"),
		KeepStageFile: keepStageFile,
		Noop:          config.Noop,
		Prefix:        config.Backend[0].Prefix,
		SyncOnly:      config.SyncOnly,
		TemplateDir:   filepath.Join(config.ConfDir, "templates"),
	}
	return nil
}

func getBackendNodesFromSRV(record, scheme string) ([]string, error) {
	nodes := make([]string, 0)

	// Ignore the CNAME as we don't need it.
	_, addrs, err := net.LookupSRV("", "", record)
	if err != nil {
		return nodes, err
	}
	for _, srv := range addrs {
		host := strings.TrimRight(srv.Target, ".")
		port := strconv.FormatUint(uint64(srv.Port), 10)
		nodes = append(nodes, fmt.Sprintf("%s://%s", scheme, net.JoinHostPort(host, port)))
	}
	return nodes, nil
}

// processFlags iterates through each flag set on the command line and
// overrides corresponding configuration settings.
func processFlags() {
	flag.Visit(setConfigFromFlag)
}

func processEnv() {
	cakeys := os.Getenv("CONFD_CLIENT_CAKEYS")
	if len(cakeys) > 0 {
		config.Backend[0].ClientCaKeys = cakeys
	}

	cert := os.Getenv("CONFD_CLIENT_CERT")
	if len(cert) > 0 {
		config.Backend[0].ClientCert = cert
	}

	key := os.Getenv("CONFD_CLIENT_KEY")
	if len(key) > 0 {
		config.Backend[0].ClientKey = key
	}
}

func setConfigFromFlag(f *flag.Flag) {
	switch f.Name {
	case "auth-token":
		config.Backend[0].AuthToken = authToken
	case "auth-type":
		config.Backend[0].AuthType = authType
	case "backend":
		config.Backend[0].Type = backend
	case "basic-auth":
		config.Backend[0].BasicAuth = basicAuth
	case "client-cert":
		config.Backend[0].ClientCert = clientCert
	case "client-key":
		config.Backend[0].ClientKey = clientKey
	case "client-ca-keys":
		config.Backend[0].ClientCaKeys = clientCaKeys
	case "confdir":
		config.ConfDir = confdir
	case "node":
		config.Backend[0].BackendNodes = nodes
	case "interval":
		config.Interval = interval
	case "noop":
		config.Noop = noop
	case "password":
		config.Backend[0].Password = password
	case "prefix":
		config.Backend[0].Prefix = prefix
	case "scheme":
		config.Backend[0].Scheme = scheme
	case "srv-domain":
		config.Backend[0].SRVDomain = srvDomain
	case "srv-record":
		config.Backend[0].SRVRecord = srvRecord
	case "sync-only":
		config.SyncOnly = syncOnly
	case "table":
		config.Backend[0].Table = table
	case "username":
		config.Backend[0].Username = username
	case "log-level":
		config.LogLevel = logLevel
	case "watch":
		config.Watch = watch
	case "app-id":
		config.Backend[0].AppID = appID
	case "user-id":
		config.Backend[0].UserID = userID
	}
}
